ch_7_Finetune/7_1_Finetune_Llama2-7B.md

```diff
@@ -0,0 +1,321 @@
+tokenizer.padding_side = "left"
+```
+> **Note**
+>
+> If you have already downloaded the Llama 2 (7B) model, you could specify `pretrained_model_name_or_path` to the local model path.
+
+### 7.1.2.5 Run the Training
+
+You can then start the training process by setting the `trainer` with existing tools on the HF ecosystem. Here we set `warmup_steps` to be 20 to accelerate the process of training.
+```python
+import transformers
+trainer = transformers.Trainer(
+    model=model,
+    train_dataset=data["train"],
+    args=transformers.TrainingArguments(
+        per_device_train_batch_size=4,
+        gradient_accumulation_steps= 1,
+        warmup_steps=20,
+        max_steps=200,
+        learning_rate=2e-4,
+        save_steps=100,
+        fp16=True,
+        logging_steps=20,
+        output_dir="outputs", # specify your own output path here
+        optim="adamw_hf", # paged_adamw_8bit is not supported yet
+        # gradient_checkpointing=True, # can further reduce memory but slower
+    ),
+    data_collator=transformers.DataCollatorForLanguageModeling(tokenizer, mlm=False),
+)
+model.config.use_cache = False  # silence the warnings, and we should re-enable it for inference
+result = trainer.train()
+print(result)
+```
+We can get the following outputs showcasing our training loss:
+```
+/home/arda/anaconda3/envs/yining-llm-qlora/lib/python3.9/site-packages/transformers/optimization.py:411: FutureWarning: This implementation of AdamW is deprecated and will be removed in a future version. Use the PyTorch implementation torch.optim.AdamW instead, or set `no_deprecation_warning=True` to disable this warning
+  warnings.warn(
+{'loss': 1.7193, 'learning_rate': 0.0002, 'epoch': 0.03}                                                             
+{'loss': 1.3242, 'learning_rate': 0.00017777777777777779, 'epoch': 0.06}                                             
+{'loss': 1.2266, 'learning_rate': 0.00015555555555555556, 'epoch': 0.1}                                              
+{'loss': 1.1534, 'learning_rate': 0.00013333333333333334, 'epoch': 0.13}                                             
+{'loss': 0.9368, 'learning_rate': 0.00011111111111111112, 'epoch': 0.16}                                             
+{'loss': 0.9321, 'learning_rate': 8.888888888888889e-05, 'epoch': 0.19}                                              
+{'loss': 0.9902, 'learning_rate': 6.666666666666667e-05, 'epoch': 0.22}                                              
+{'loss': 0.8593, 'learning_rate': 4.4444444444444447e-05, 'epoch': 0.26}                                             
+{'loss': 1.0055, 'learning_rate': 2.2222222222222223e-05, 'epoch': 0.29}                                             
+{'loss': 1.0081, 'learning_rate': 0.0, 'epoch': 0.32}                                                                
+{'train_runtime': 285.5368, 'train_samples_per_second': 2.802, 'train_steps_per_second': 0.7, 'train_loss': 1.1155566596984863, 'epoch': 0.32}
```

Comment:
Should we hide the specific value of `train_runtime`, `train_samples_per_second` and `train_steps_per_second`